cluster-autoscaler/core/scaledown/actuation/priority_evictor.go

```diff
@@ -0,0 +1,238 @@
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+
+	context "k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/core/scaledown/status"
+	"k8s.io/autoscaler/cluster-autoscaler/metrics"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/daemonset"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	pod_util "k8s.io/autoscaler/cluster-autoscaler/utils/pod"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	kubelet_config "k8s.io/kubernetes/pkg/kubelet/apis/config"
+)
+
+// PriorityEvictor can be used to evict pods from nodes.
+type PriorityEvictor struct {
+	EvictionRetryTime                time.Duration
+	DsEvictionRetryTime              time.Duration
+	DsEvictionEmptyNodeTimeout       time.Duration
+	PodEvictionHeadroom              time.Duration
+	evictionRegister                 evictionRegister
+	deleteOptions                    simulator.NodeDeleteOptions
+	shutdownGracePeriodByPodPriority []kubelet_config.ShutdownGracePeriodByPodPriority
+}
+
+// NewPriorityEvictor returns an instance of PriorityEvictor using the default parameters.
+func NewPriorityEvictor(deleteOptions simulator.NodeDeleteOptions, evictionRegister evictionRegister, shutdownGracePeriodByPodPriority []kubelet_config.ShutdownGracePeriodByPodPriority) PriorityEvictor {
+	sort.Slice(shutdownGracePeriodByPodPriority, func(i, j int) bool {
+		return shutdownGracePeriodByPodPriority[i].Priority < shutdownGracePeriodByPodPriority[j].Priority
+	})
+
+	return PriorityEvictor{
+		EvictionRetryTime:                DefaultEvictionRetryTime,
+		DsEvictionRetryTime:              DefaultDsEvictionRetryTime,
+		DsEvictionEmptyNodeTimeout:       DefaultDsEvictionEmptyNodeTimeout,
+		PodEvictionHeadroom:              DefaultPodEvictionHeadroom,
+		evictionRegister:                 evictionRegister,
+		deleteOptions:                    deleteOptions,
+		shutdownGracePeriodByPodPriority: shutdownGracePeriodByPodPriority,
+	}
+}
+
+// DrainNode groups pods in the node in to priority groups and, evicts pods in the asending order of priorities.
+// All non-mirror, non-DS pods on the node are evicted. Mirror pods are not evicted.
+// DaemonSet pods are evicted if DaemonSetEvictionForOccupiedNodes is enabled, or if they have the EnableDsEvictionKey annotation.
+func (e PriorityEvictor) DrainNode(ctx *context.AutoscalingContext, nodeInfo *framework.NodeInfo) (map[string]status.PodEvictionResult, error) {
+	pods := podsToEvictCombined(ctx, nodeInfo)
+	return e.DrainNodeWithPodsBasedOnPodPriority(ctx, nodeInfo.Node(), pods)
+}
+
+// EvictDaemonSetPods creates eviction objects for all DaemonSet pods on the node.
+func (e PriorityEvictor) EvictDaemonSetPods(ctx *context.AutoscalingContext, nodeInfo *framework.NodeInfo, timeNow time.Time) error {
+	return evictDaemonSetPods(ctx, nodeInfo, timeNow, e.deleteOptions, e.DsEvictionEmptyNodeTimeout, e.DsEvictionRetryTime, e.evictionRegister)
```

Comment:
Do we need to bother about the eviction order when only evicting DaemonSet pods?

Reply:
Yes, definitely. During scale-down, this function is called for nodes with only DS pods, while `DrainNode` is called for nodes with non-DS pods. In the current evictor, the behavior for DS pods in both these functions is more or less the same - initiate the evictions, and don't wait for the pods to disappear. We should keep the behavior the same for the new evictor as well - if `DrainNode` looks at priorities, so should the other one. From a use-case perspective, some DS pods depending on other DS pods during termination is not that much more surprising than regular pods depending on DS pods.

I actually don't think we need `EvictDaemonSetPods` at all, since the only thing that makes sense is for it to work the same as `DrainNode` anyway (for DS pods). We could just call `DrainNode` for both empty and non-empty nodes in scale-down. WDYT?

Reply:
Done